service/src/test/java/bio/terra/pipelines/service/PipelineRunsServiceTest.java

```diff
@@ -0,0 +1,262 @@
+  }
+
+  @Test
+  void testGetCorrectNumberOfRows() {
+    // A test row should exist for this user.
+    List<PipelineRun> pipelineRuns = pipelineRunsRepository.findAllByUserId(testUserId);
+    assertEquals(1, pipelineRuns.size());
+
+    // insert another row and verify that it shows up
+    PipelineRun newPipelineRun = createTestJobWithJobId(testJobId);
+
+    pipelineRunsRepository.save(newPipelineRun);
+    pipelineRuns = pipelineRunsRepository.findAllByUserId(testUserId);
+    assertEquals(2, pipelineRuns.size());
+  }
+
+  @Test
+  void testCorrectUserIsolation() {
+    // A test row should exist for this user.
+    List<PipelineRun> pipelineRuns = pipelineRunsRepository.findAllByUserId(testUserId);
+    assertEquals(1, pipelineRuns.size());
+
+    // insert row for second user and verify that it shows up
+    String testUserId2 = TestUtils.TEST_USER_ID_2;
+    PipelineRun newPipelineRun = createTestJobWithJobIdAndUser(UUID.randomUUID(), testUserId2);
+    pipelineRunsRepository.save(newPipelineRun);
+
+    // Verify that the old userid still show only 1 record
+    pipelineRuns = pipelineRunsRepository.findAllByUserId(testUserId);
+    assertEquals(1, pipelineRuns.size());
+
+    // Verify the new user's id shows a single job as well
+    pipelineRuns = pipelineRunsRepository.findAllByUserId(testUserId2);
+    assertEquals(1, pipelineRuns.size());
+  }
+
+  @Test
+  void createPipelineRunNoWorkspaceSetUp() {
+    Pipeline testPipelineWithIdMissingWorkspaceId = createTestPipelineWithId();
+    testPipelineWithIdMissingWorkspaceId.setWorkspaceId(null);
+
+    assertThrows(
+        InternalServerErrorException.class,
+        () ->
+            pipelineRunsService.createPipelineRun(
+                testPipelineWithIdMissingWorkspaceId,
+                testJobId,
+                testUserId,
+                testDescription,
+                testPipelineInputs,
+                testResultUrl));
+  }
+
+  @Test
+  void createPipelineRunImputation() {
+    Pipeline testPipelineWithId = createTestPipelineWithId();
+
+    // override this mock to ensure the correct flight class is being requested
+    when(mockJobBuilder.flightClass(RunImputationJobFlight.class)).thenReturn(mockJobBuilder);
+
+    PipelineRun returnedPipelineRun =
+        pipelineRunsService.createPipelineRun(
+            testPipelineWithId,
+            testJobId,
+            testUserId,
+            testDescription,
+            testPipelineInputs,
+            testResultUrl);
+
+    assertEquals(testJobId, returnedPipelineRun.getJobId());
+    assertEquals(testUserId, returnedPipelineRun.getUserId());
+    assertEquals(testDescription, returnedPipelineRun.getDescription());
+    assertEquals(testResultUrl, returnedPipelineRun.getResultUrl());
+    assertEquals(testPipelineWithId.getId(), returnedPipelineRun.getPipelineId());
+    //    assertNotNull(returnedPipelineRun.getCreated());
```

Comment:
these checks fail - I think because we're not out of the transaction at this point so it hasn't actually done the write and generated the timestamps in the db?

Reply:
it fails here but the `findByJobIdAndUserId` call below returns a pipeline that does ahve a created and updated time? hmmm i wonder if there are annotations we can add to those columns in the PipelineRunEntity so it would make this test work here